src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,337 @@
+                if (IsCandidateSymbol(memberReference.Member))
+                {
+                    // Get the value usage info.
+                    var valueUsageInfo = memberReference.GetValueUsageInfo();
+
+                    if (valueUsageInfo == ValueUsageInfo.ReadWrite)
+                    {
+                        Debug.Assert(memberReference.Parent is ICompoundAssignmentOperation compoundAssignment &&
+                            compoundAssignment.Target == memberReference ||
+                            memberReference.Parent is IIncrementOrDecrementOperation);
+
+                        // Compound assignment or increment whose value is being dropped (parent has null type)
+                        // is treated as a Write as the value was never actually 'read' in a way that is observable.
+                        if (memberReference.Parent.Parent?.Type == null)
+                        {
+                            valueUsageInfo = ValueUsageInfo.Write;
+                        }
+                    }
+
+                    OnSymbolUsage(memberReference.Member, valueUsageInfo);
+                }
+            }
+
+            public void AnalyzeInvocationOperation(OperationAnalysisContext operationContext)
+            {
+                var invocation = (IInvocationOperation)operationContext.Operation;
+                if (IsCandidateSymbol(invocation.TargetMethod))
+                {
+                    OnSymbolUsage(invocation.TargetMethod, ValueUsageInfo.Read);
+                }
+            }
+
+            public void OnSymbolEnd(SymbolAnalysisContext symbolEndContext, bool hasInvalidOperation)
+            {
+                if (hasInvalidOperation)
+                {
+                    return;
+                }
+
+                // Report diagnostics for unused candidate members.
+                ImmutableHashSet<ISymbol> symbolsReferencedInDocComments = null;
+                var members = ((INamedTypeSymbol)symbolEndContext.Symbol).GetMembers();
+                foreach (var member in members)
+                {
+                    // Check if the underlying member is neither read nor a readable reference to the member is taken.
+                    if (TryRemove(member, out var valueUsageInfo) &&
+                        !valueUsageInfo.ContainsReadOrReadableRef())
+                    {
+                        Debug.Assert(IsCandidateSymbol(member));
+                        Debug.Assert(!member.IsImplicitlyDeclared);
+
+                        if (symbolsReferencedInDocComments == null)
+                        {
+                            // Bail out if there are syntax errors in any of the declarations of the containing type.
+                            if (HasSyntaxErrors())
+                            {
+                                return;
+                            }
+
+                            symbolsReferencedInDocComments = GetCandidateSymbolsReferencedInDocComments();
```

Comment:
i find this logic exceptionally hard to read.  presumably, it's depending on implicit local state capture to make sense.  but that makes it super hard to assess.  As an example:

1. my brain immediately says: why are you checking HasSyntaxErrors each time through the loop?

The answer, presumably, is that HasSyntaxErrors implicitly depends on 'member', and thus could result in a different answer each time.  But that is not at all clear from the code or the name.  Either:

1. do not use implicit local capture to manipulate flow control (my *strong* preference)
2. name the member to make it clear that it depends on implicit capture. i.e. "CurrentMemberHasSyntaxErrors()

Please do '1'.  For my sanity :)